File: backend/opencode/codex.go
package opencode

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/enough/enough/backend/auth"
)

func codexRequestHeaders(accessToken string) map[string]string {
	return auth.CodexCloudflareHeaders(accessToken)
}

type responsesRequest struct {
	Model        string              `json:"model"`
	Instructions string              `json:"instructions"`
	Input        any                 `json:"input"`
	Tools        any                 `json:"tools,omitempty"`
	Store        bool                `json:"store"`
	Stream       bool                `json:"stream,omitempty"`
	Reasoning    *responsesReasoning `json:"reasoning,omitempty"`
}

type responsesReasoning struct {
	Effort  string `json:"effort,omitempty"`
	Summary string `json:"summary,omitempty"`
}

type responsesResponse struct {
	Status     string           `json:"status"`
	Output     []responsesItem  `json:"output"`
	OutputText string           `json:"output_text,omitempty"`
	Error      *responsesError  `json:"error,omitempty"`
	Usage      *responsesUsage  `json:"usage,omitempty"`
}

type responsesError struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

type responsesUsage struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
	TotalTokens  int `json:"total_tokens"`
}

type responsesItem struct {
	Type      string `json:"type"`
	Role      string `json:"role,omitempty"`
	Status    string `json:"status,omitempty"`
	Name      string `json:"name,omitempty"`
	CallID    string `json:"call_id,omitempty"`
	Arguments string `json:"arguments,omitempty"`
	ID        string `json:"id,omitempty"`
	Content   []struct {
		Type string `json:"type"`
		Text string `json:"text"`
	} `json:"content,omitempty"`
	Summary []struct {
		Text string `json:"text"`
	} `json:"summary,omitempty"`
}

func (c *Client) responsesURL() string {
	base := strings.TrimRight(c.baseURL, "/")
	if strings.HasSuffix(base, "/responses") {
		return base
	}
	return base + "/responses"
}

func (c *Client) buildResponsesRequest(req ChatRequest) (responsesRequest, error) {
	if req.Model == "" {
		req.Model = c.model
	}
	msgs := PrepareRequestMessages(req.Messages, req.Model)
	instructions, convMsgs := splitInstructionsMessages(msgs)
	if strings.TrimSpace(instructions) == "" {
		instructions = "You are a helpful coding assistant."
	}
	input := messagesToResponsesInput(convMsgs)
	tools := chatToolsToResponses(req.Tools)

	out := responsesRequest{
		Model:        req.Model,
		Instructions: instructions,
		Input:        input,
		Tools:        tools,
		Store:        false,
		Reasoning:    reasoningFromChatRequest(req),
	}
	return out, nil
}

func splitInstructionsMessages(msgs []Message) (instructions string, rest []Message) {
	var parts []string
	for _, msg := range msgs {
		switch msg.Role {
		case "system", "developer":
			if text := strings.TrimSpace(ContentString(msg)); text != "" {
				parts = append(parts, text)
			}
		default:
			rest = append(rest, msg)
		}
	}
	return strings.Join(parts, "\n\n"), rest
}

func reasoningFromChatRequest(req ChatRequest) *responsesReasoning {
	if req.Thinking != nil && req.Thinking.Type == "disabled" {
		return nil
	}
	effort := req.ReasoningEffort
	if effort == "" {
		return nil
	}
	if effort == "max" {
		effort = "xhigh"
	}
	return &responsesReasoning{Effort: effort, Summary: "auto"}
}

func (c *Client) chatResponsesOnce(ctx context.Context, req ChatRequest) (Message, error) {
	bodyReq, err := c.buildResponsesRequest(req)
	if err != nil {
		return Message{}, err
	}

	body, err := json.Marshal(bodyReq)
	if err != nil {
		return Message{}, err
	}

	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.responsesURL(), bytes.NewReader(body))
	if err != nil {
		return Message{}, err
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Authorization", "Bearer "+c.apiKey)
	for k, v := range codexRequestHeaders(c.apiKey) {
		httpReq.Header.Set(k, v)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return Message{}, err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return Message{}, err
	}
	if resp.StatusCode >= 400 {
		return Message{}, fmt.Errorf("codex %d: %s", resp.StatusCode, strings.TrimSpace(string(raw)))
	}

	var out responsesResponse
	if err := json.Unmarshal(raw, &out); err != nil {
		return Message{}, fmt.Errorf("decode codex response: %w", err)
	}

	return parseResponsesMessage(out)
}

func messagesToResponsesInput(msgs []Message) []any {
	var items []any
	for _, msg := range msgs {
		switch msg.Role {
		case "system":
			continue
		case "user":
			blocks := ContentBlocks(msg)
			hasImage := false
			for _, b := range blocks {
				if b.Type == "image_url" {
					hasImage = true
					break
				}
			}
			if hasImage {
				var content []map[string]any
				for _, b := range blocks {
					switch b.Type {
					case "text":
						if b.Text != "" {
							content = append(content, map[string]any{"type": "input_text", "text": b.Text})
						}
					case "image_url":
						if b.ImageURL != nil {
							content = append(content, map[string]any{
								"type":      "input_image",
								"image_url": b.ImageURL.URL,
								"detail":    "auto",
							})
						}
					}
				}
				items = append(items, map[string]any{
					"role":    "user",
					"content": content,
				})
			} else {
				text := ContentString(msg)
				items = append(items, map[string]any{
					"role":    "user",
					"content": text,
				})
			}
		case "assistant":
			text := ContentString(msg)
			if text != "" {
				items = append(items, map[string]any{
					"role": "assistant",
					"content": []map[string]string{
						{"type": "output_text", "text": text},
					},
				})
			}
			for _, tc := range msg.ToolCalls {
				callID := tc.ID
				if callID == "" {
					callID = fmt.Sprintf("call_%d", len(items))
				}
				items = append(items, map[string]any{
					"type":      "function_call",
					"call_id":   callID,
					"name":      tc.Function.Name,
					"arguments": tc.Function.Arguments,
				})
			}
		case "tool":
			callID := msg.ToolCallID
			if callID == "" {
				continue
			}
			blocks := ContentBlocks(msg)
			hasImage := false
			for _, b := range blocks {
				if b.Type == "image_url" {
					hasImage = true
					break
				}
			}
			if hasImage {
				var outputItems []map[string]any
				for _, b := range blocks {
					if b.Type == "image_url" && b.ImageURL != nil {
						outputItems = append(outputItems, map[string]any{
							"type":      "input_image",
							"image_url": b.ImageURL.URL,
						})
					} else {
						outputItems = append(outputItems, map[string]any{
							"type": "input_text",
							"text": b.Text,
						})
					}
				}
				items = append(items, map[string]any{
					"type":    "function_call_output",
					"call_id": callID,
					"output":  outputItems,
				})
			} else {
				items = append(items, map[string]any{
					"type":    "function_call_output",
					"call_id": callID,
					"output":  ContentString(msg),
				})
			}
		}
	}
	return items
}

func chatToolsToResponses(tools []Tool) any {
	if len(tools) == 0 {
		return nil
	}
	out := make([]map[string]any, 0, len(tools))
	for _, t := range tools {
		if t.Function.Name == "" {
			continue
		}
		params := t.Function.Parameters
		if len(params) == 0 {
			params = json.RawMessage(`{"type":"object","properties":{}}`)
		}
		out = append(out, map[string]any{
			"type":        "function",
			"name":        t.Function.Name,
			"description": t.Function.Description,
			"strict":      false,
			"parameters":  json.RawMessage(params),
		})
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

func parseResponsesMessage(resp responsesResponse) (Message, error) {
	status := strings.ToLower(strings.TrimSpace(resp.Status))
	if status == "failed" || status == "cancelled" {
		msg := "codex request failed"
		if resp.Error != nil && resp.Error.Message != "" {
			msg = resp.Error.Message
		}
		return Message{}, fmt.Errorf("codex: %s", msg)
	}

	msg := Message{Role: "assistant"}
	if resp.Usage != nil {
		msg.Usage = &Usage{
			Input:  resp.Usage.InputTokens,
			Output: resp.Usage.OutputTokens,
			TotalTokens: resp.Usage.TotalTokens,
		}
	}

	output := resp.Output
	if len(output) == 0 && strings.TrimSpace(resp.OutputText) != "" {
		s := strings.TrimSpace(resp.OutputText)
		msg.Content = StringContent(s)
		return msg, nil
	}
	if len(output) == 0 {
		return Message{}, fmt.Errorf("codex: empty response")
	}

	var textParts []string
	var reasoningParts []string
	for _, item := range output {
		switch item.Type {
		case "message":
			for _, part := range item.Content {
				if part.Type == "output_text" || part.Type == "text" {
					if part.Text != "" {
						textParts = append(textParts, part.Text)
					}
				}
			}
		case "reasoning":
			for _, part := range item.Summary {
				if part.Text != "" {
					reasoningParts = append(reasoningParts, part.Text)
				}
			}
		case "function_call":
			if item.Status != "" && item.Status != "completed" {
				continue
			}
			callID := item.CallID
			if callID == "" && strings.HasPrefix(item.ID, "fc_") {
				callID = "call_" + strings.TrimPrefix(item.ID, "fc_")
			}
			if callID == "" {
				callID = fmt.Sprintf("call_%d", len(msg.ToolCalls))
			}
			args := item.Arguments
			if args == "" {
				args = "{}"
			}
			msg.ToolCalls = append(msg.ToolCalls, ToolCall{
				ID:   callID,
				Type: "function",
				Function: ToolCallFunction{
					Name:      item.Name,
					Arguments: args,
				},
			})
		}
	}

	if len(textParts) > 0 {
		s := strings.Join(textParts, "")
		msg.Content = StringContent(s)
	}
	if len(reasoningParts) > 0 {
		s := strings.Join(reasoningParts, "\n")
		msg.ReasoningContent = &s
	}

	if len(msg.Content) == 0 && len(msg.ToolCalls) == 0 {
		return Message{}, fmt.Errorf("codex: no assistant output")
	}
	return msg, nil
}

File: backend/opencode/codex.ts
// PORT: backend/opencode/codex.go

import { Effect } from "effect";
import { codex_cloudflare_headers } from "../auth/codex_headers";
import { client, client_error } from "./client";
import { prepare_request_messages } from "./messages";
import {
  blocks_content,
  content_blocks,
  content_string,
  get_reasoning,
  string_content,
  type chat_request,
  type message,
  type tool,
  type tool_call,
  type usage
} from "./types";

// OpenAI Responses API Types
export type responses_reasoning = {
  effort?: string;
  summary?: string;
};

export type responses_request = {
  model: string;
  instructions: string;
  input: unknown;
  tools?: unknown;
  store: boolean;
  stream?: boolean;
  reasoning?: responses_reasoning;
};

export type responses_error = {
  code: string;
  message: string;
};

export type responses_usage = {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
};

export type responses_item = {
  type: string;
  role?: string;
  status?: string;
  name?: string;
  call_id?: string;
  arguments?: string;
  id?: string;
  content?: {
    type: string;
    text: string;
  }[];
  summary?: {
    text: string;
  }[];
};

export type responses_response = {
  status: string;
  output: responses_item[];
  output_text?: string;
  error?: responses_error;
  usage?: responses_usage;
};

export const codex_request_headers = (access_token: string): Record<string, string> => {
  return codex_cloudflare_headers(access_token);
};

export const responses_url = (c: client): string => {
  const base = c.base_url.replace(/\/+$/, "");
  if (base.endsWith("/responses")) {
    return base;
  }
  return base + "/responses";
};

export const split_instructions_messages = (msgs: message[]): [string, message[]] => {
  const parts: string[] = [];
  const rest: message[] = [];
  for (const msg of msgs) {
    if (msg.role === "system" || msg.role === "developer") {
      const text = content_string(msg).trim();
      if (text !== "") {
        parts.push(text);
      }
    } else {
      rest.push(msg);
    }
  }
  return [parts.join("\n\n"), rest];
};

export const reasoning_from_chat_request = (req: chat_request): responses_reasoning | undefined => {
  if (req.thinking !== undefined && req.thinking !== null && (req.thinking as { type?: string }).type === "disabled") {
    return undefined;
  }
  let effort = req.reasoning_effort ?? "";
  if (effort === "") {
    return undefined;
  }
  if (effort === "max") {
    effort = "xhigh";
  }
  return { effort, summary: "auto" };
};

export const messages_to_responses_input = (msgs: message[]): unknown[] => {
  const items: unknown[] = [];
  for (const msg of msgs) {
    switch (msg.role) {
      case "system":
        continue;
      case "user": {
        const blocks = content_blocks(msg);
        let has_image = false;
        if (blocks !== null) {
          for (const b of blocks) {
            if (b.type === "image_url") {
              has_image = true;
              break;
            }
          }
        }
        if (has_image && blocks !== null) {
          const content: unknown[] = [];
          for (const b of blocks) {
            switch (b.type) {
              case "text":
                if (b.text !== undefined && b.text !== "") {
                  content.push({ type: "input_text", text: b.text });
                }
                break;
              case "image_url":
                if (b.image_url !== undefined && b.image_url !== null) {
                  content.push({
                    type: "input_image",
                    image_url: b.image_url.url,
                    detail: "auto",
                  });
                }
                break;
            }
          }
          items.push({ role: "user", content });
        } else {
          const text = content_string(msg);
          items.push({ role: "user", content: text });
        }
        break;
      }
      case "assistant": {
        const text = content_string(msg);
        if (text !== "") {
          items.push({
            role: "assistant",
            content: [{ type: "output_text", text }],
          });
        }
        if (msg.tool_calls !== undefined) {
          for (const tc of msg.tool_calls) {
            let call_id = tc.id;
            if (call_id === "") {
              call_id = `call_${items.length}`;
            }
            items.push({
              type: "function_call",
              call_id,
              name: tc.function.name,
              arguments: tc.function.arguments,
            });
          }
        }
        break;
      }
      case "tool": {
        const call_id = msg.tool_call_id;
        if (call_id === undefined || call_id === "") {
          continue;
        }
        const blocks = content_blocks(msg);
        let has_image = false;
        if (blocks !== null) {
          for (const b of blocks) {
            if (b.type === "image_url") {
              has_image = true;
              break;
            }
          }
        }
        if (has_image && blocks !== null) {
          const output_items: unknown[] = [];
          for (const b of blocks) {
            if (b.type === "image_url" && b.image_url !== undefined && b.image_url !== null) {
              output_items.push({
                type: "input_image",
                image_url: b.image_url.url,
              });
            } else {
              output_items.push({
                type: "input_text",
                text: b.text ?? "",
              });
            }
          }
          items.push({
            type: "function_call_output",
            call_id,
            output: output_items,
          });
        } else {
          items.push({
            type: "function_call_output",
            call_id,
            output: content_string(msg),
          });
        }
        break;
      }
    }
  }
  return items;
};

export const chat_tools_to_responses = (tools?: tool[]): unknown => {
  if (tools === undefined || tools.length === 0) {
    return undefined;
  }
  const out: unknown[] = [];
  for (const t of tools) {
    if (t.function.name === "") {
      continue;
    }
    let params: unknown = { type: "object", properties: {} };
    if (t.function.parameters !== undefined && t.function.parameters.length > 0) {
      try {
        params = JSON.parse(new TextDecoder().decode(t.function.parameters));
      } catch {}
    }
    out.push({
      type: "function",
      name: t.function.name,
      description: t.function.description,
      strict: false,
      parameters: params,
    });
  }
  if (out.length === 0) {
    return undefined;
  }
  return out;
};

export const build_responses_request = (c: client, req: chat_request): responses_request => {
  if (req.model === "") {
    req.model = c.model;
  }
  const msgs = prepare_request_messages(req.messages, req.model);
  let [instructions, conv_msgs] = split_instructions_messages(msgs);
  if (instructions.trim() === "") {
    instructions = "You are a helpful coding assistant.";
  }
  const input = messages_to_responses_input(conv_msgs);
  const tools = chat_tools_to_responses(req.tools);

  return {
    model: req.model,
    instructions,
    input,
    tools,
    store: false,
    reasoning: reasoning_from_chat_request(req),
  };
};

export const parse_responses_message = (resp: responses_response): message => {
  const status = (resp.status ?? "").toLowerCase().trim();
  if (status === "failed" || status === "cancelled") {
    let msg = "codex request failed";
    if (resp.error?.message !== undefined && resp.error.message !== "") {
      msg = resp.error.message;
    }
    throw new Error(`codex: ${msg}`);
  }

  let usage_data: usage | undefined = undefined;
  if (resp.usage !== undefined && resp.usage !== null) {
    usage_data = {
      input: resp.usage.input_tokens,
      output: resp.usage.output_tokens,
      totalTokens: resp.usage.total_tokens,
    };
  }

  const msg: message = {
    role: "assistant",
    content: new Uint8Array(),
    usage: usage_data,
  };

  const output = resp.output ?? [];
  if (output.length === 0 && resp.output_text !== undefined && resp.output_text.trim() !== "") {
    msg.content = string_content(resp.output_text.trim());
    return msg;
  }
  if (output.length === 0) {
    throw new Error("codex: empty response");
  }

  const text_parts: string[] = [];
  const reasoning_parts: string[] = [];
  for (const item of output) {
    switch (item.type) {
      case "message":
        if (item.content !== undefined) {
          for (const part of item.content) {
            if (part.type === "output_text" || part.type === "text") {
              if (part.text !== "") {
                text_parts.push(part.text);
              }
            }
          }
        }
        break;
      case "reasoning":
        if (item.summary !== undefined) {
          for (const part of item.summary) {
            if (part.text !== "") {
              reasoning_parts.push(part.text);
            }
          }
        }
        break;
      case "function_call": {
        if (item.status !== undefined && item.status !== "" && item.status !== "completed") {
          continue;
        }
        let call_id = item.call_id ?? "";
        if (call_id === "" && item.id !== undefined && item.id.startsWith("fc_")) {
          call_id = "call_" + item.id.slice("fc_".length);
        }
        if (call_id === "") {
          const count = msg.tool_calls?.length ?? 0;
          call_id = `call_${count}`;
        }
        let args = item.arguments ?? "";
        if (args === "") {
          args = "{}";
        }
        if (msg.tool_calls === undefined) {
          msg.tool_calls = [];
        }
        msg.tool_calls.push({
          id: call_id,
          type: "function",
          function: {
            name: item.name ?? "",
            arguments: args,
          },
        });
        break;
      }
    }
  }

  if (text_parts.length > 0) {
    msg.content = string_content(text_parts.join(""));
  }
  if (reasoning_parts.length > 0) {
    msg.reasoning_content = reasoning_parts.join("\n");
  }

  if (msg.content.length === 0 && (msg.tool_calls === undefined || msg.tool_calls.length === 0)) {
    throw new Error("codex: no assistant output");
  }

  return msg;
};

export const chat_responses_once = (
  c: client,
  ctx: AbortSignal,
  req: chat_request
): Effect.Effect<message, client_error> => {
  return Effect.tryPromise({
    try: async () => {
      const body_req = build_responses_request(c, req);
      const body = JSON.stringify(body_req);

      const headers: Record<string, string> = {
        "Content-Type": "application/json",
        Authorization: `Bearer ${c.api_key}`,
        ...codex_request_headers(c.api_key),
      };

      const resp = await fetch(responses_url(c), {
        method: "POST",
        signal: ctx,
        headers,
        body,
      });

      const raw = await resp.text();
      if (resp.status >= 400) {
        throw new Error(`codex ${resp.status}: ${raw.trim()}`);
      }

      const out = JSON.parse(raw) as responses_response;
      return parse_responses_message(out);
    },
    catch: (cause) => client_error("chat_responses_once", cause),
  });
};

/*
PORT STATUS
source path: backend/opencode/codex.go
source lines: 398
draft lines: 382
confidence: high
status: phase_b_compile
*/
